File: Python3/Sliding Window/1004. Max Consecutive Ones III.py
""" 1004. Max Consecutive Ones III 
Difficulty: Medium
Submission Link: https://leetcode.com/problems/max-consecutive-ones-iii/submissions/1391225751/ """

class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        l = 0
        max_num = 0

        for r in range(len(nums)):
            k -= 1 - nums[r]

            if k < 0:
                k += 1 - nums[l]
                l += 1
            else:
                max_num = max(max_num, r - l + 1)

        return max_num

File: C/Sliding Window/1004. Max Consecutive Ones III.c
/* 1004. Max Consecutive Ones III 
Difficulty: Medium
Submission Link: https://leetcode.com/problems/max-consecutive-ones-iii/submissions/1391230063/ */

int longestOnes(int* nums, int numsSize, int k) {
    int zero_cnt = 0, l = 0, ans = 0;
    for (int r = 0; r < numsSize; ++r) {
        if (nums[r] == 0)
            zero_cnt++;
        while (zero_cnt > k)
            if (nums[l++] == 0)
                zero_cnt--;
        ans = fmax(ans, r - l + 1);
    }
    return ans;
}